File: Tests/ConnectorControlCoreTests/MasterStoreTests.swift
import XCTest
import ConnectorControlTestSupport
@testable import ConnectorControlCore

final class MasterStoreTests: XCTestCase {
    var tempDir: TempDir!
    var dir: URL!
    var url: URL { dir.appendingPathComponent("mcps.json") }

    override func setUpWithError() throws {
        tempDir = TempDir(prefix: "store")
        dir = tempDir.url
    }

    override func tearDownWithError() throws {
        tempDir.dispose()
    }

    func testEnabledServersRendersEnabledSubset() {
        let store = MasterStore.single([
            "on": MCPEntry(enabled: true, config: .object(["command": .string("a")])),
            "off": MCPEntry(enabled: false, config: .object(["command": .string("b")]))])
        XCTAssertEqual(store.enabledServers, ["on": .object(["command": .string("a")])])
    }

    func testLoadMissingFileReturnsEmptyStore() {
        let result = MasterStoreIO.load(from: url)
        XCTAssertEqual(result.store, .empty)
        XCTAssertNil(result.corruptFileURL)
    }

    func testSaveThenLoadRoundTrips() throws {
        var store = MasterStore.empty
        store.mcps["scoutbook"] = MCPEntry(
            enabled: false,
            config: RemotePattern.make(url: "https://example.com/mcp"),
            lastEditView: .json)
        try MasterStoreIO.save(store, to: url)
        let result = MasterStoreIO.load(from: url)
        XCTAssertEqual(result.store, store)
        XCTAssertNil(result.corruptFileURL)
    }

    func testLoadCorruptFilePreservesItAndReturnsEmpty() throws {
        try Data("{not json!!".utf8).write(to: url)
        let result = MasterStoreIO.load(from: url)
        XCTAssertEqual(result.store, .empty)
        let corrupt = try XCTUnwrap(result.corruptFileURL)
        XCTAssertTrue(corrupt.lastPathComponent.hasPrefix("mcps.corrupt."))
        XCTAssertEqual(try String(contentsOf: corrupt, encoding: .utf8), "{not json!!")
        XCTAssertFalse(FileManager.default.fileExists(atPath: url.path))
    }

    func testReadIsSideEffectFree() throws {
        // Missing file → nil, nothing created.
        XCTAssertNil(MasterStoreIO.read(from: url))
        // Corrupt file → nil, file left exactly in place (unlike load, which
        // moves it aside — the watcher must be able to peek at a sync tool's
        // mid-write partial without destroying it).
        try Data("{not json!!".utf8).write(to: url)
        XCTAssertNil(MasterStoreIO.read(from: url))
        XCTAssertEqual(try String(contentsOf: url, encoding: .utf8), "{not json!!")
        // Valid file → decoded store.
        var store = MasterStore.empty
        store.mcps["s"] = MCPEntry(config: .object(["command": .string("npx")]))
        try MasterStoreIO.save(store, to: url)
        XCTAssertEqual(MasterStoreIO.read(from: url), store)
    }

    func testBackupTimestampFormat() throws {
        var components = DateComponents()
        components.year = 2025
        components.month = 7
        components.day = 15
        components.hour = 17
        components.minute = 20
        components.second = 0
        components.nanosecond = 123_000_000
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        let date = try XCTUnwrap(calendar.date(from: components))
        XCTAssertEqual(BackupTimestamp.string(from: date), "2025-07-15T17-20-00-123Z")
    }

    func testBackupTimestampsSortChronologicallyAcrossDSTFallBack() {
        // 2026-11-01 America/New_York repeats 01:00–02:00 wall-clock; UTC
        // stamps must stay strictly increasing regardless.
        let start = Date(timeIntervalSince1970: 1_793_500_000)  // hours before
        var previous = ""
        for step in 0..<10 {
            let stamp = BackupTimestamp.string(
                from: start.addingTimeInterval(Double(step) * 1800))
            XCTAssertGreaterThan(stamp, previous)
            previous = stamp
        }
    }

    func testUnknownActiveProfileFallsBackToExistingProfile() throws {
        let json = """
        {"version":2,"activeProfile":"Ghost",\
        "profiles":{"Alpha":{"mcps":{}},"Beta":{"mcps":{}}}}
        """
        try Data(json.utf8).write(to: url)
        let result = MasterStoreIO.load(from: url)
        XCTAssertNil(result.corruptFileURL)
        XCTAssertEqual(result.store.activeProfile, "Alpha", "sorted-first existing profile")
    }

    func testV1FormatFileIsTreatedAsCorruptAndRebuilt() throws {
        // No v1 compatibility: an old-format file can't decode against the
        // v2-only schema, so it flows through the existing corrupt-file path
        // (moved aside, empty store returned) rather than being migrated.
        let json = """
        {"version":1,"mcps":{"scoutbook":{"enabled":true,\
        "config":{"command":"npx","args":["-y","mcp-remote","https://example.com/mcp"]},\
        "lastEditView":"form"}}}
        """
        try Data(json.utf8).write(to: url)
        let result = MasterStoreIO.load(from: url)
        XCTAssertEqual(result.store, .empty)
        let corrupt = try XCTUnwrap(result.corruptFileURL)
        XCTAssertTrue(corrupt.lastPathComponent.hasPrefix("mcps.corrupt."))
        XCTAssertFalse(FileManager.default.fileExists(atPath: url.path))
    }

    /// MasterStore has no custom `init(from:)` — decoding is exactly as
    /// strict as the synthesized Codable conformance, so every one of these
    /// six malformed shapes fails to decode and flows through the same
    /// corrupt-file path as garbage bytes: moved aside, empty store returned.
    func testRequiredKeysAreStrictLikeCodable() throws {
        let malformed = [
            // missing "version"
            #"{"activeProfile":"Default","profiles":{"Default":{"mcps":{}}}}"#,
            // missing "activeProfile"
            #"{"version":2,"profiles":{"Default":{"mcps":{}}}}"#,
            // missing "profiles"
            #"{"version":2,"activeProfile":"Default"}"#,
            // "version" is a string, not an Int
            #"{"version":"2","activeProfile":"Default","profiles":{"Default":{"mcps":{}}}}"#,
            // "profiles" is an array, not an object
            #"{"version":2,"activeProfile":"Default","profiles":[]}"#,
            // an entry missing its required "config"
            #"{"version":2,"activeProfile":"Default","profiles":{"Default":{"mcps":{"a":{"enabled":true,"lastEditView":"form"}}}}}"#,
        ]
        for (index, json) in malformed.enumerated() {
            try Data(json.utf8).write(to: url)
            // A distinct `now` per case: two loads in the same millisecond would
            // otherwise collide on the same corrupt-file name and prove nothing.
            let result = MasterStoreIO.load(from: url, now: Date(timeIntervalSince1970: 1_800_000_000 + Double(index)))
            XCTAssertEqual(result.store, .empty, "case \(index): \(json)")
            let corrupt = try XCTUnwrap(result.corruptFileURL, "case \(index): \(json)")
            XCTAssertTrue(corrupt.lastPathComponent.hasPrefix("mcps.corrupt."), "case \(index)")
        }
    }

    func testUnknownKeysAreIgnored() throws {
        let json = """
        {"version":2,"activeProfile":"Default",\
        "profiles":{"Default":{"mcps":{}}},"unknownField":"surprise"}
        """
        try Data(json.utf8).write(to: url)
        let result = MasterStoreIO.load(from: url)
        XCTAssertNil(result.corruptFileURL)
        XCTAssertEqual(result.store, .empty)
    }

    /// MasterStore.version is a Swift Int (64-bit), unlike a 32-bit field:
    /// a value past Int32.max still decodes.
    func testVersionBeyondInt32IsAccepted() throws {
        let json = """
        {"version":5000000000,"activeProfile":"Default",\
        "profiles":{"Default":{"mcps":{}}}}
        """
        try Data(json.utf8).write(to: url)
        let result = MasterStoreIO.load(from: url)
        XCTAssertNil(result.corruptFileURL)
        XCTAssertEqual(result.store.version, 5_000_000_000)
    }

    func testLoadCorruptFileReportsOriginalPathWhenMoveFails() throws {
        let fixedNow = Date(timeIntervalSince1970: 1_752_600_000)
        let garbage = "{not json!!"
        try Data(garbage.utf8).write(to: url)

        // Pre-create the aside file so moveItem will fail due to name collision.
        let stamp = BackupTimestamp.string(from: fixedNow)
        let aside = url.deletingLastPathComponent()
            .appendingPathComponent("mcps.corrupt.\(stamp).json")
        try Data("existing".utf8).write(to: aside)

        let result = MasterStoreIO.load(from: url, now: fixedNow)
        XCTAssertEqual(result.store, .empty)
        XCTAssertEqual(result.corruptFileURL, url)
        XCTAssertEqual(try String(contentsOf: url, encoding: .utf8), garbage)
    }
}

File: windows/tests/ConnectorControl.Core.Tests/MasterStoreTests.cs
using System.Text;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests;

/// <summary>Mirror: Tests/ConnectorControlCoreTests/MasterStoreTests.swift</summary>
public class MasterStoreTests : IDisposable
{
    private readonly TempDir dir = new("store");
    private string Url => dir.File("mcps.json");

    public void Dispose() => dir.Dispose();

    private static JsonValue Cmd(string command) => JsonValue.Object(("command", JsonValue.String(command)));

    [Fact]
    public void EnabledServersRendersEnabledSubset()
    {
        var store = new MasterStore(new Dictionary<string, McpEntry>
        {
            ["on"] = new McpEntry(true, Cmd("a")),
            ["off"] = new McpEntry(false, Cmd("b")),
        });
        Assert.Equal(new Dictionary<string, JsonValue> { ["on"] = Cmd("a") }, store.EnabledServers);
    }

    [Fact]
    public void RenamesANonActiveCollectionAndKeepsTheActiveName()
    {
        var store = new MasterStore(2, "A", [
            new KeyValuePair<string, Collection>("A", new Collection()),
            new KeyValuePair<string, Collection>("B", new Collection()),
        ]);
        Assert.Null(store.RenameCollection("B", "C"));
        Assert.Equal(new HashSet<string>(["A", "C"], StringComparer.Ordinal), store.Collections.Keys.ToHashSet(StringComparer.Ordinal));
        Assert.Equal("A", store.ActiveCollection);
        Assert.Null(store.RenameCollection("A", "Z"));
        Assert.Equal("Z", store.ActiveCollection);
        Assert.NotNull(store.RenameCollection("Z", "C"));   // a taken name is refused
        Assert.Equal("No collection named “Nope”.", store.RenameCollection("Nope", "Q"));
        Assert.Equal("No collection named “Nope”.", store.DeleteCollection("Nope"));
    }

    [Fact]
    public void DeletesANonActiveCollectionAndRefusesTheLast()
    {
        var store = new MasterStore(2, "A", [
            new KeyValuePair<string, Collection>("A", new Collection()),
            new KeyValuePair<string, Collection>("B", new Collection()),
        ]);
        Assert.Null(store.DeleteCollection("B"));
        Assert.NotNull(store.DeleteCollection("A"));   // the last collection stays
        store = new MasterStore(2, "B", [
            new KeyValuePair<string, Collection>("A", new Collection()),
            new KeyValuePair<string, Collection>("B", new Collection()),
        ]);
        Assert.Null(store.DeleteCollection("B"));
        Assert.Equal("A", store.ActiveCollection);   // deleting the active one moves to the sorted-first remaining
    }

    [Fact]
    public void LoadMissingFileReturnsEmptyStore()
    {
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Equal(MasterStore.Empty(), store);
        Assert.Null(corrupt);
    }

    /// <summary>C#-only: a missing collection is looked up, not optional-chained, so the read has to be shown to create nothing.</summary>
    [Fact]
    public void ReadingMcpsDoesNotCreateACollection()
    {
        var store = new MasterStore(2, "Work", [new KeyValuePair<string, Collection>("Work", new Collection())]);
        store.ActiveCollection = "Ghost";   // an active collection the store has no Collection object for
        Assert.Empty(store.Mcps);
        Assert.False(store.Collections.ContainsKey("Ghost"), "reading Mcps must not create a collection as a side effect");
        Assert.Single(store.Collections);
    }

    /// <summary>C#-only: the Mac counts through EnabledServers and has no separate count to pin.</summary>
    [Fact]
    public void EnabledCountCountsWithoutBuildingTheServerDictionary()
    {
        var store = new MasterStore(new Dictionary<string, McpEntry>
        {
            ["on"] = new McpEntry(true, Cmd("a")),
            ["off"] = new McpEntry(false, Cmd("b")),
            ["on2"] = new McpEntry(true, Cmd("c")),
        });
        Assert.Equal(2, store.EnabledCount);
    }

    [Fact]
    public void SaveThenLoadRoundTrips()
    {
        var store = MasterStore.Empty();
        store.Mcps["scoutbook"] = new McpEntry(false,
            JsonValue.Object(("command", JsonValue.String("npx")),
                ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("mcp-remote"), JsonValue.String("https://example.com/mcp")]))),
            EditView.Json);
        MasterStoreIO.Save(store, Url);
        var (loaded, corrupt) = MasterStoreIO.Load(Url);
        Assert.Equal(store, loaded);
        Assert.Null(corrupt);
    }

    [Fact]
    public void LoadCorruptFilePreservesItAndReturnsEmpty()
    {
        File.WriteAllText(Url, "{not json!!");
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Equal(MasterStore.Empty(), store);
        Assert.NotNull(corrupt);
        Assert.StartsWith("mcps.corrupt.", Path.GetFileName(corrupt), StringComparison.Ordinal);
        Assert.Equal("{not json!!", File.ReadAllText(corrupt));
        Assert.False(File.Exists(Url));
    }

    [Fact]
    public void ReadIsSideEffectFree()
    {
        Assert.Null(MasterStoreIO.Read(Url));
        File.WriteAllText(Url, "{not json!!");
        Assert.Null(MasterStoreIO.Read(Url));
        Assert.Equal("{not json!!", File.ReadAllText(Url));
        var store = MasterStore.Empty();
        store.Mcps["s"] = new McpEntry(JsonValue.Object(("command", JsonValue.String("npx"))));
        MasterStoreIO.Save(store, Url);
        Assert.Equal(store, MasterStoreIO.Read(Url));
    }

    [Fact]
    public void UnknownActiveCollectionFallsBackToExistingCollection()
    {
        File.WriteAllText(Url, """
            {"version":2,"activeProfile":"Ghost","profiles":{"Alpha":{"mcps":{}},"Beta":{"mcps":{}}}}
            """);
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Null(corrupt);
        Assert.Equal("Alpha", store.ActiveCollection);   // sorted-first existing collection
    }

    /// <summary>
    /// No v1 compatibility: an old-format file can't decode against the v2-only schema, so it flows
    /// through the existing corrupt-file path (moved aside, empty store returned) rather than being
    /// migrated.
    /// </summary>
    [Fact]
    public void V1FormatFileIsTreatedAsCorruptAndRebuilt()
    {
        File.WriteAllText(Url, "{\"version\":1,\"mcps\":{\"scoutbook\":{\"enabled\":true,\"config\":{\"command\":\"npx\",\"args\":[\"-y\",\"mcp-remote\",\"https://example.com/mcp\"]},\"lastEditView\":\"form\"}}}");
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Equal(MasterStore.Empty(), store);
        Assert.NotNull(corrupt);
        Assert.StartsWith("mcps.corrupt.", Path.GetFileName(corrupt), StringComparison.Ordinal);
        Assert.False(File.Exists(Url));
    }

    /// <summary>
    /// Decoding is exactly as strict as Swift's synthesized Codable conformance, so every one of
    /// these malformed shapes fails to decode, as it does on the Mac.
    /// </summary>
    [Theory]
    [InlineData("{\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{}}}}")]                        // missing "version"
    [InlineData("{\"version\":2,\"profiles\":{\"Default\":{\"mcps\":{}}}}")]                                        // missing "activeProfile"
    [InlineData("{\"version\":2,\"activeProfile\":\"Default\"}")]                                                     // missing "profiles"
    [InlineData("{\"version\":\"2\",\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{}}}}")]         // "version" is a string
    [InlineData("{\"version\":2,\"activeProfile\":\"Default\",\"profiles\":[]}")]                                    // "profiles" is an array
    [InlineData("{\"version\":2,\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{\"a\":{\"enabled\":true,\"lastEditView\":\"form\"}}}}}")]  // an entry missing "config"
    [InlineData("{\"version\":2,\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{\"a\":{\"enabled\":true,\"config\":{}}}}}}")]           // an entry missing "lastEditView"
    [InlineData("{\"version\":2,\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{\"a\":{\"enabled\":\"yes\",\"config\":{},\"lastEditView\":\"form\"}}}}}")]  // "enabled" is not a bool
    [InlineData("{\"version\":2,\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{\"a\":{\"enabled\":true,\"config\":{},\"lastEditView\":\"grid\"}}}}}")]  // an unknown view
    [InlineData("[]")]                                                                                                     // not an object at all
    public void RequiredKeysAreStrictLikeCodable(string json)
    {
        Assert.Throws<FormatException>(() => MasterStore.FromJson(JsonValue.Parse(json)));
    }

    [Fact]
    public void UnknownKeysAreIgnored()
    {
        var store = MasterStore.FromJson(JsonValue.Parse(
            "{\"version\":2,\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{},\"note\":\"x\"}},\"unknownField\":\"surprise\"}"));
        Assert.Equal(MasterStore.Empty(), store);
    }

    /// <summary>MasterStore.Version is a long, as Swift's Int is 64-bit: a value past Int32.MaxValue still decodes.</summary>
    [Fact]
    public void VersionBeyondInt32IsAccepted()
    {
        var store = MasterStore.FromJson(JsonValue.Parse("{\"version\":5000000000,\"activeProfile\":\"Default\",\"profiles\":{\"Default\":{\"mcps\":{}}}}"));
        Assert.Equal(5_000_000_000L, store.Version);
    }

    [Fact]
    public void LoadCorruptFileReportsOriginalPathWhenMoveFails()
    {
        var fixedNow = DateTime.UnixEpoch.AddSeconds(1_752_600_000);
        File.WriteAllText(Url, "{not json!!");
        var aside = dir.File($"mcps.corrupt.{BackupTimestamp.From(fixedNow)}.json");
        File.WriteAllText(aside, "existing");
        var (store, corrupt) = MasterStoreIO.Load(Url, fixedNow);
        Assert.Equal(MasterStore.Empty(), store);
        Assert.Equal(Url, corrupt);
        Assert.Equal("{not json!!", File.ReadAllText(Url));
    }

    [Fact]
    public void SavedFileUsesAppleEncoderFormat()
    {
        var store = MasterStore.Empty();
        store.Mcps["s"] = new McpEntry(JsonValue.Object(("command", JsonValue.String("npx")), ("args", JsonValue.Array([JsonValue.String("https://x.y/z")]))));
        MasterStoreIO.Save(store, Url);
        const string expected =
            "{\n  \"activeProfile\" : \"Default\",\n  \"profiles\" : {\n    \"Default\" : {\n      \"mcps\" : {\n        \"s\" : {\n" +
            "          \"config\" : {\n            \"args\" : [\n              \"https:\\/\\/x.y\\/z\"\n            ],\n            \"command\" : \"npx\"\n          },\n" +
            "          \"enabled\" : true,\n          \"lastEditView\" : \"form\"\n        }\n      }\n    }\n  },\n  \"version\" : 2\n}";
        Assert.Equal(expected, File.ReadAllText(Url, Encoding.UTF8));
    }
}
